### PR TITLE
[UpdateTestChecks] Add support for SPIRV in update_llc_test_checks.py

### PR DESCRIPTION
**Pull request by pmatos**
We have to change the location of the comment for function start
in SPIRV. We add a new command line flag --skip-check-label, because
there's no function label on SPIRV.

Previously https://reviews.llvm.org/D157858

### CODE REVIEW

File: llvm/utils/UpdateTestChecks/common.py [hunk 1]
```diff
@@ -196,6 +196,13 @@ def __call__(self, parser, namespace, values, option_string=None):
         default=[],
         help="List of regular expressions such that, for matching global value declarations, literal integer values should be encoded in hex in the associated FileCheck directives",
     )
+    parser.add_argument(
+        "--skip-check-label",
+        action="store_true",
+        dest="skip_check_label",
+        default=False,
+        help="Activate CHECK-LABEL line generation per function",
```
Comment (RoboTux) [inline llvm/utils/UpdateTestChecks/common.py:204]: Shouldn't it be "disable CHECK-LABEL" line generation?

**Comment by arichardson** [comment]
Could we just emit a different label format that matches the comment for spirv? Having to pass the flag seems a bit awkward

File: llvm/utils/UpdateTestChecks/common.py (hunk 1 above)
Comment (pmatos) [reply llvm/utils/UpdateTestChecks/common.py:204]: That's correct. Thanks.

**Comment by pmatos** [comment]
> Could we just emit a different label format that matches the comment for spirv? Having to pass the flag seems a bit awkward

I would love to be able to do that but the `check_label_format` string is sufficiently hardcoded that changing this feels like it could be potentially a pandora's box:
https://github.com/llvm/llvm-project/blob/7f3610ac69191a2a445fa9b4b947019a2fd51fb1/llvm/utils/UpdateTestChecks/asm.py#L585

Happy to hear any ideas that might ease changing this without breaking other stuff.

**Comment by RKSimon** [comment]
How different would check_label_format have to be? Surely SPIRV has something similar that we can use?

I always like to see at least one use of the update script when we add a new target - please can you regenerate a suitable file in CodeGen\SPIRV to show the generated checks?

**Comment by pmatos** [comment]
> How different would check_label_format have to be? Surely SPIRV has something similar that we can use?
> 

The assembly for the test I just committed looks like:

```
	OpCapability Kernel
	OpCapability Addresses
	OpCapability Linkage
	OpCapability OptNoneINTEL
	OpExtension "SPV_INTEL_optnone"
	%1 = OpExtInstImport "OpenCL.std"
	OpMemoryModel Physical32 OpenCL
	OpExecutionMode %4 ContractionOff
	OpSource Unknown 0
	OpName %4 "_Z3foov"
	OpDecorate %4 LinkageAttributes "_Z3foov" Export
	%2 = OpTypeVoid
	%3 = OpTypeFunction %2
                                        ; -- Begin function _Z3foov
	%4 = OpFunction %2 DontInline %3
	%5 = OpLabel
	OpReturn
	OpFunctionEnd
                                       ; -- End function
```

Besides the comment there's no obvious place where the function body starts and its name due to the structure of the SPIRV file.

> I always like to see at least one use of the update script when we add a new target - please can you regenerate a suitable file in CodeGen\SPIRV to show the generated checks?

Now updated the commit to include one of these.

**Comment by pmatos** [comment]
> Why not just match OpName ? ASM_FUNCTION_X86_RE does something similar in matching the func name a few lines before the "Begin function" tag

As far as I understand we need to grab the body of the function. In the case of SPIRV, the body doesn't come after the name. Take the case of 2 functions called void1 and void2:

```
	OpCapability Kernel
	OpCapability Addresses
	OpCapability Linkage
	OpCapability OptNoneINTEL
	OpExtension "SPV_INTEL_optnone"
	%1 = OpExtInstImport "OpenCL.std"
	OpMemoryModel Physical32 OpenCL
	OpExecutionMode %4 ContractionOff
	OpExecutionMode %5 ContractionOff
	OpSource Unknown 0
	OpName %4 "void1"
	OpName %5 "void2"
	OpDecorate %4 LinkageAttributes "void1" Export
	OpDecorate %5 LinkageAttributes "void2" Export
	%2 = OpTypeVoid
	%3 = OpTypeFunction %2
	%4 = OpFunction %2 DontInline %3        ; -- Begin function void1
	%6 = OpLabel
	OpReturn
	OpFunctionEnd
                                        ; -- End function
	%5 = OpFunction %2 DontInline %3        ; -- Begin function void2
	%7 = OpLabel
	OpReturn
	OpFunctionEnd
                                        ; -- End function
```

How would you grab with a regex the body of void2 if I parse the name from OpName?

**Comment by RoboTux** [comment]
> > Why not just match OpName ? ASM_FUNCTION_X86_RE does something similar in matching the func name a few lines before the "Begin function" tag
> 
> As far as I understand we need to grab the body of the function. In the case of SPIRV, the body doesn't come after the name. Take the case of 2 functions called void1 and void2:
> 
> ```
> 	OpCapability Kernel
> 	OpCapability Addresses
> 	OpCapability Linkage
> 	OpCapability OptNoneINTEL
> 	OpExtension "SPV_INTEL_optnone"
> 	%1 = OpExtInstImport "OpenCL.std"
> 	OpMemoryModel Physical32 OpenCL
> 	OpExecutionMode %4 ContractionOff
> 	OpExecutionMode %5 ContractionOff
> 	OpSource Unknown 0
> 	OpName %4 "void1"
> 	OpName %5 "void2"
> 	OpDecorate %4 LinkageAttributes "void1" Export
> 	OpDecorate %5 LinkageAttributes "void2" Export
> 	%2 = OpTypeVoid
> 	%3 = OpTypeFunction %2
> 	%4 = OpFunction %2 DontInline %3        ; -- Begin function void1
> 	%6 = OpLabel
> 	OpReturn
> 	OpFunctionEnd
>                                         ; -- End function
> 	%5 = OpFunction %2 DontInline %3        ; -- Begin function void2
> 	%7 = OpLabel
> 	OpReturn
> 	OpFunctionEnd
>                                         ; -- End function
> ```
> 
> How would you grab with a regex the body of void2 if I parse the name from OpName?

How about something like something along the line of:
"OpName (?P<func>%[0-9_]+)(.*\n)+"\1 = OpFunction.*\n(?P<body>.*?)OpFunctionEnd"

**Comment by arichardson** [comment]
Maybe something like `"[ \t]+; \-\- Begin function (?P<func>[^\n]+)(?P<func_name_separator>)\n(?P<body>.*?)\n[ \t]+; \-\- End function"` would work? By setting func_name_separator to an empty match the CHECK-LABEL string expansion won't add the default value of ":"? Is that sufficient?
```
                    check_label_format
                    % (
                        checkprefix,
                        funcdef_attrs_and_ret,
                        func_name,
                        args_and_sig,
                        func_name_separator,
                    )
```

**Comment by pmatos** [comment]
> > > Why not just match OpName ? ASM_FUNCTION_X86_RE does something similar in matching the func name a few lines before the "Begin function" tag
> > 
> > 
> > As far as I understand we need to grab the body of the function. In the case of SPIRV, the body doesn't come after the name. Take the case of 2 functions called void1 and void2:
> > ```
> > 	OpCapability Kernel
> > 	OpCapability Addresses
> > 	OpCapability Linkage
> > 	OpCapability OptNoneINTEL
> > 	OpExtension "SPV_INTEL_optnone"
> > 	%1 = OpExtInstImport "OpenCL.std"
> > 	OpMemoryModel Physical32 OpenCL
> > 	OpExecutionMode %4 ContractionOff
> > 	OpExecutionMode %5 ContractionOff
> > 	OpSource Unknown 0
> > 	OpName %4 "void1"
> > 	OpName %5 "void2"
> > 	OpDecorate %4 LinkageAttributes "void1" Export
> > 	OpDecorate %5 LinkageAttributes "void2" Export
> > 	%2 = OpTypeVoid
> > 	%3 = OpTypeFunction %2
> > 	%4 = OpFunction %2 DontInline %3        ; -- Begin function void1
> > 	%6 = OpLabel
> > 	OpReturn
> > 	OpFunctionEnd
> >                                         ; -- End function
> > 	%5 = OpFunction %2 DontInline %3        ; -- Begin function void2
> > 	%7 = OpLabel
> > 	OpReturn
> > 	OpFunctionEnd
> >                                         ; -- End function
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > How would you grab with a regex the body of void2 if I parse the name from OpName?
> 
> How about something like something along the line of: "OpName (?P%[0-9_]+)(._\n)+"\1 = OpFunction._\n(?P.*?)OpFunctionEnd"

Thanks, I have updated the patch with a regex based on your suggestion that seems to work well.

**Comment by pmatos** [comment]
> Maybe something like `"[ \t]+; \-\- Begin function (?P<func>[^\n]+)(?P<func_name_separator>)\n(?P<body>.*?)\n[ \t]+; \-\- End function"` would work? By setting func_name_separator to an empty match the CHECK-LABEL string expansion won't add the default value of ":"? Is that sufficient?
> 
> ```
>                     check_label_format
>                     % (
>                         checkprefix,
>                         funcdef_attrs_and_ret,
>                         func_name,
>                         args_and_sig,
>                         func_name_separator,
>                     )
> ```

To be honest, I don't understand the suggestion. I removed - for explanatory purposes - the flag from the commit and regenerated the test. Now the check lines look like:

```
; CHECK-EXTENSION-LABEL: _Z3foov:
; CHECK-EXTENSION:       %4 = OpFunction %2 DontInline %3
; CHECK-EXTENSION-NEXT:    %5 = OpLabel
; CHECK-EXTENSION-NEXT:    OpReturn
; CHECK-EXTENSION-NEXT:    OpFunctionEnd
```

The problem is the *-LABEL line which is not issues at all by SPIRV. It seems you know how to skip this line without having an extra flag but I don't understand what you mean by: 

> the CHECK-LABEL string expansion won't add the default value of ":"?

By guess at an answer is that this is still not sufficient because there's no label at all to match.

File: llvm/test/CodeGen/SPIRV/extensions/SPV_INTEL_optnone.ll [hunk 2]
```diff
@@ -1,3 +1,4 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 3
 ; RUN: llc -O0 -mtriple=spirv32-unknown-unknown --spirv-extensions=SPV_INTEL_optnone %s -o - | FileCheck %s --check-prefix=CHECK-EXTENSION
 ; RUN: llc -O0 -mtriple=spirv32-unknown-unknown %s -o - | FileCheck %s --check-prefix=CHECK-NO-EXTENSION
```
Comment (RKSimon) [inline llvm/test/CodeGen/SPIRV/extensions/SPV_INTEL_optnone.ll:3]: Can you add a common prefix to ensure its working correctly?
```
; RUN: llc -O0 -mtriple=spirv32-unknown-unknown --spirv-extensions=SPV_INTEL_optnone %s -o - | FileCheck %s --check-prefixes=CHECK,CHECK-EXTENSION
; RUN: llc -O0 -mtriple=spirv32-unknown-unknown %s -o - | FileCheck %s --check-prefixes=CHECK,CHECK-NO-EXTENSION
```

**Review by RKSimon — APPROVED** [review]
LGTM - cheers

**Comment by pmatos** [comment]
> The ':' is added because that's the default value of `func_name_separator` if it's not set by the regex. I was suggesting to include an empty capture in the regex so that `func_name_separator` is set to the empty string instead.

Ah yes, I understand what you are saying and I see you have in your suggested RE you have `(?P<func_name_separator>)` but that will still result in:
```
; CHECK-LABEL: _Z3foov
; CHECK:       %4 = OpFunction %2 DontInline %3
; CHECK-NEXT:    %5 = OpLabel
; CHECK-NEXT:    OpReturn
; CHECK-NEXT:    OpFunctionEnd
```

but SPIRV issues no labels. The odd thing (for me) is that:

* the test passes even with a `CHECK-LABEL: _Z3foov` even if there's no label;
* The last lines say `These prefixes are unused...` even if they are used early on.

**Comment by RKSimon** [comment]
>     * The last lines say `These prefixes are unused...` even if they are used early on.

That's just a sideeffect of some of the prefixes being manually added

**Comment by pmatos** [comment]
> > ```
> > * The last lines say `These prefixes are unused...` even if they are used early on.
> > ```
> 
> That's just a sideeffect of some of the prefixes being manually added

Thanks - I will merge this then.

**Pull request merged**